group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java

```diff
@@ -135,56 +148,71 @@ private GroupAssignment newAssignmentHomogeneous(
-        // in the new assignment and have members in current assignment by step 2. But we haven't implemented it to avoid the
-        // complexity in both the implementation and the run time complexity. This step was mentioned in the KIP to reduce
-        // the burden of certain members of the share groups. This can be achieved with the help of limiting the max
-        // no. of partitions assignment for every member(KAFKA-18788). Hence, the potential problem of burdening
-        // the share consumers will be addressed in a future PR.
+        // For entirely balanced assignment, we would expect (numTargetPartitions / numGroupMembers) partitions per member, rounded upwards.
+        // That can be expressed as         Math.ceil(numTargetPartitions / (double) numGroupMembers)
+        // Using integer arithmetic, as     (numTargetPartitions + numGroupMembers - 1) / numGroupMembers
+        int numGroupMembers = groupSpec.memberIds().size();
+        int numTargetPartitions = targetPartitions.size();
+        int desiredAssignmentCount = (numTargetPartitions + numGroupMembers - 1) / numGroupMembers;
+
+        Map<TopicIdPartition, List<String>> newAssignment = new HashMap<>((int) (((numTargetPartitions + 1) / 0.75f) + 1));
+
+        // Hash member IDs to topic partitions. Each member will be assigned one partition, but some partitions
+        // might have been assigned to more than one member.
+        memberHashAssignment(groupSpec.memberIds(), targetPartitions, newAssignment);
+
+        // Combine current and new hashed assignments, sized to accommodate the expected number of mappings.
+        Map<String, Set<TopicIdPartition>> finalAssignment = new HashMap<>((int) ((numGroupMembers / 0.75f) + 1));
+        Map<TopicIdPartition, Set<String>> finalAssignmentByPartition = new HashMap<>((int) (((numTargetPartitions + 1) / 0.75f) + 1));
+
+        // First, take the members assigned by hashing.
+        newAssignment.forEach((targetPartition, members) -> members.forEach(member -> {
+            finalAssignment.computeIfAbsent(member, k -> new HashSet<>()).add(targetPartition);
+            finalAssignmentByPartition.computeIfAbsent(targetPartition, k -> new HashSet<>()).add(member);
+        }));
 
-        newAssignment.forEach((targetPartition, members) -> members.forEach(member ->
-            finalAssignment.computeIfAbsent(member, k -> new HashSet<>()).add(targetPartition)));
+        // Then, take the members from the current assignment, making sure that no member has too many assigned partitions.
         // When combining current assignment, we need to only consider the topics in current assignment that are also being
         // subscribed in the new assignment as well.
         currentAssignment.forEach((targetPartition, members) -> {
-            if (subscribedTopicIds.contains(targetPartition.topicId()))
+            if (subscribedTopicIds.contains(targetPartition.topicId())) {
```

Comment:
I want to double check my understanding.

Here we start with the current assignment and remove any topics which are no longer subscribed and remove any members which aren't currently around. We then copy assignments from currentAssignment into finalAssignment up to the desired number of assignments per member.

This leaves us with a subset of currentAssignment and possibly a number of unassigned partitions. 

We then round-robbin the remaining unassigned partitions into currentAssignment (up to the desired count).

Reply:
That's the idea.